typed_python/PythonSerializationContext.hpp

```diff
@@ -655,6 +713,105 @@ class PythonSerializationContext : public SerializationContext {
 
 
 private:
+    template<class Size_Fn, class GetItem_Fn>
+    inline void serializeIndexable(PyObject* o, SerializationBuffer& b, Size_Fn size_fn, GetItem_Fn get_item_fn) const {
+        uint32_t id;
+        bool isNew;
+        std::tie(id, isNew) = b.cachePointer(o);
+        b.write_uint32(id);
+
+        if (isNew) {
+            size_t sz = size_fn(o);
+
+            b.write_uint32(sz);
+
+            for (long k = 0; k < sz; k++) {
+                serializePythonObject(get_item_fn(o, k), b);
+            }
+        }
+    }
+
+    template<class Factory_Fn, class SetItem_Fn>
+    inline PyObject* deserializeIndexable(DeserializationBuffer& b, Factory_Fn factory_fn, SetItem_Fn set_item_fn) const {
+        uint32_t id = b.read_uint32();
+
+        PyObject* res = (PyObject*)b.lookupCachedPointer(id);
+        if (res) {
+            return incref(res);
+        }
+
+        size_t sz = b.read_uint32();
+
+        res = factory_fn(sz);
+        b.addCachedPointer(id, incref(res), true);
+
+        try {
+            for (long k = 0; k < sz; k++) {
+                set_item_fn(res, k, deserializePythonObject(b));
+            }
+        } catch(...) {
+            Py_DECREF(res);
+            throw;
+        }
+
+        return res;
+    }
+
+    template<class Size_Fn>
+    inline void serializeIterable(PyObject* o, SerializationBuffer& b, Size_Fn size_fn) const {
+        uint32_t id;
+        bool isNew;
+        std::tie(id, isNew) = b.cachePointer(o);
+        b.write_uint32(id);
+
+        if (isNew) {
+            size_t sz = size_fn(o);
+
+            b.write_uint32(sz);
+            PyObject* iter = PyObject_GetIter(o);
+            if (!iter) {
+                return;
+            }
+            PyObject* item = PyIter_Next(iter);
+            while (item) {
+                serializePythonObject(item, b);
+                Py_DECREF(item);
+                item = PyIter_Next(iter);
+            }
+            Py_DECREF(iter);
+        }
+    }
+
+    template<class Factory_Fn, class AddItem_Fn>
+    inline PyObject* deserializeIterable(DeserializationBuffer &b, Factory_Fn factory_fn, AddItem_Fn add_item_fn) const {
+        uint32_t id = b.read_uint32();
+
+        PyObject* res = (PyObject*)b.lookupCachedPointer(id);
+        if (res) {
+            return incref(res);
+        }
+
+        size_t sz = b.read_uint32();
+
+        res = factory_fn(NULL);
+        b.addCachedPointer(id, incref(res), true);
+
+        try {
+            for (long k = 0; k < sz; k++) {
+                int success = add_item_fn(res, deserializePythonObject(b));
```

Comment:
are we decreffing the result of 'deserializePythonObject' anywhere?

Reply:
Good question! So `add_item_fn` is `PySet_Add`. It's documentation does not specify if it steals the reference to the key being added to the set. Theoretically it should to keep symmetry with `PyList_SetItem` and `PyTuple_SetItem`.

Reply:
Baaaah!! Looking at the code it looks like `PySet_Add` does not steal the reference. It does `Py_INCREF(key)`. Bogus!